src/Mod/Sketcher/Gui/SnapManager.h

```diff
@@ -0,0 +1,117 @@
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ *   SnapManager initially funded by the Open Toolchain Foundation         *
+ ***************************************************************************/
+
+#ifndef SKETCHERGUI_SnapManager_H
+#define SKETCHERGUI_SnapManager_H
+
+
+#include <App/Application.h>
+
+
+namespace SketcherGui {
+
+class ViewProviderSketch;
+
+
+class ViewProviderSketchSnapAttorney {
+private:
+
+    static inline int getPreselectPoint(const ViewProviderSketch& vp);
+    static inline int getPreselectCross(const ViewProviderSketch& vp);
+    static inline int getPreselectCurve(const ViewProviderSketch& vp);
+
+    friend class SnapManager;
+};
+
+/* This class is used to manage the overriding of mouse pointer coordinates in Sketcher
+*  (in Edit-Mode) depending on the situation. Those situations are in priority order :
+*  1 - Snap at angle: For tools like Slot, Arc, Line, Ellipse, this enables to constrain the angle at steps of 5� (or customized angle).
+*  This is useful to make features at a certain angle (45� for example)
+*  2 - Snap to object: This snaps the mouse pointer onto objects.
+*  3 - Snap to grid: This snaps the mouse pointer on the grid.
+*/
+class SnapManager
+{
+
+    /** @brief      Class for monitoring changes in parameters affecting Snapping
+    *  @details
+    *
+    * This nested class is a helper responsible for attaching to the parameters relevant for
+    * SnapManager, initialising the SnapManager to the current configuration
+    * and handle in real time any change to their values.
+    */
+    class ParameterObserver : public ParameterGrp::ObserverType
+    {
+    public:
+        explicit ParameterObserver(SnapManager& client);
+        ~ParameterObserver() override;
+
+        void subscribeToParameters();
+
+        void unsubscribeToParameters();
+
+        /** Observer for parameter group. */
+        void OnChange(Base::Subject<const char*>& rCaller, const char* sReason) override;
+
+    private:
+        void initParameters();
+        void updateSnapToObjectParameter(const std::string& parametername);
+        void updateSnapAngleParameters(const std::string& parametername);
+
+    private:
+        std::map<std::string, std::function<void(const std::string&)>> str2updatefunction;
+        SnapManager& Client;
+    };
+
+public:
+    explicit SnapManager(ViewProviderSketch &vp);
+    ~SnapManager();
+
+    void snap(double& x, double& y);
+    void snapAtAngle(double& x, double& y);
+    bool snapToObject(double& x, double& y);
+    void snapToGrid(double& x, double& y);

```

Comment:
As future-proofing, shall also probably return a bool to indicate if snapping was or not successful.

Reply:
I would second it, but for slightly different reasons.

A snap to an angle may fail because there is nowhere to snap to following that angle. A snap to a Grid may fail, specially if there is no grid. A way should be provided to signal that.

For the future-proof. From SOLID, the L stands for Liskov substitution principle (Barbara Liskov). This more or less advocate to maintaining the same interface in different objects, enable to use them indistinctly (use generic code to handle them). That is not necessarily related to inheritance, although in c++ generally is. So, future-proof we could imagine to build a hierarchy where objects providing a snap inherit from an abstract class. If we use the same interface, we can just copy the code there when refactoring. But, I admit this is a little bit beyond scope and borderlining YAGNI. ;)